core/templates/services/prevent-page-unload-event.service.ts

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to handle prevent reload events.
+ */
+
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class PreventPageUnloadEventService {
+  private listenerActive: boolean;
+  constructor(private windowRef: WindowRef) {
+    this.listenerActive = false;
+  }
+
+  addListener(): void {
+    if (this.listenerActive) {
+      return;
+    }
+    this.windowRef.nativeWindow.addEventListener(
+      'beforeunload', this._preventPageUnloadEventHandler);
+    this.listenerActive = true;
+  }
+
+  removeListener(): void {
+    this.windowRef.nativeWindow.removeEventListener(
+      'beforeunload', this._preventPageUnloadEventHandler);
+    this.listenerActive = false;
+  }
+
+  private _preventPageUnloadEventHandler(
+      e: { preventDefault: () => void; returnValue: string; }
```

Comment:
This should be of type Event, right?

Reply:
yes but the returnValue throws an error saying it should be boolean

Reply:
Have you looked on web how is this handheld in other projects? :)

Reply:
DONE. It is `BeforeUnloadEvent` type [ref](https://developer.mozilla.org/en-US/docs/Web/API/BeforeUnloadEvent)